File: minCostforTickets.js
/**
 * @param {number[]} days
 * @param {number[]} costs
 * @return {number}
 */
var mincostTickets = function(days, costs) {
    let dp = new Array(days[days.length-1]+1).fill(0);
    let b;
    let c;
    for(let i = 1; i<dp.length; i++){
        if( days.includes(i)){
            //1 day
            let a = dp[i-1] + costs[0];
            //7 day
            if(i < 7){
                 b = costs[1];
            } else {
                b = dp[i-7] + costs[1];
            }
           if(i < 30){
                 c = costs[2];
            } else {
                c = dp[i-30] + costs[2];
            }

                dp[i] = Math.min(a,Math.min(b,c));

        } else {
            dp[i] = dp[i-1];
        }
    }
    
    return dp.pop();
    
};

//time complexity : O(n2)

File: MinCostForTickets.py
'''
    Time Complexity:
        O(365)

    Space Complexity:
        O(365)

    Did this code successfully run on LeetCode?:
        Yes

    Problems faced while coding this:
        None

    Approach:
        Memoization.

        On each day, choose either of the three passes and go the travelling next day.
        Find the minimum of these three.
        Cache similar subproblems.
'''

PASS_VALIDITY = [1, 7, 30]

class Solution:
    def __init__(self):
        self.costs = []
        self.days = set()
        self.memo = {}

    def mincostTickets(self, days: List[int], costs: List[int]) -> int:
        self.days = set(days)
        self.costs = costs
        return self.find_min_cost(1)

    def find_min_cost(self, day):
        if day > 365:
            return 0

        if day in self.memo:
            return self.memo[day]

        if day not in self.days:
            return self.find_min_cost(day + 1)

        min_cost = float('inf')

        for idx, cost in enumerate(self.costs):
            next_day = day + PASS_VALIDITY[idx]

            generate_cost = cost + self.find_min_cost(next_day)
            min_cost = min(min_cost, generate_cost)

        self.memo[day] = min_cost
        return self.memo[day]
